cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,710 @@
+}
+
+/**
+ * @brief Calculate global degree information for all vertices represented by current gpu
+ *
+ * Calculate local degree and perform row wise exclusive scan over all gpus in column
+ * communicator.
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @return Tuple of two device vectors. The first one contains per source edge-count encountered
+ * by gpus in the column communicator before current gpu. The second device vector contains the
+ * global out degree for every source represented by current gpu
+ */
+template <typename GraphViewType>
+std::tuple<rmm::device_uvector<typename GraphViewType::edge_type>,
+           rmm::device_uvector<typename GraphViewType::edge_type>>
+get_global_degree_information(raft::handle_t const& handle, GraphViewType const& graph_view)
+{
+  using edge_t       = typename GraphViewType::edge_type;
+  auto local_degrees = detail::calculate_local_degrees(handle, graph_view);
+
+  auto& col_comm      = handle.get_subcomm(cugraph::partition_2d::key_naming_t().col_name());
+  auto const col_size = col_comm.get_size();
+  auto const col_rank = col_comm.get_rank();
+  auto& row_comm      = handle.get_subcomm(cugraph::partition_2d::key_naming_t().row_name());
+  auto const row_size = row_comm.get_size();
+
+  auto& comm           = handle.get_comms();
+  auto const comm_size = comm.get_size();
+  auto const comm_rank = comm.get_rank();
+
+  rmm::device_uvector<edge_t> temp_input(local_degrees.size(), handle.get_stream());
+  raft::update_device(
+    temp_input.data(), local_degrees.data(), local_degrees.size(), handle.get_stream());
+
+  rmm::device_uvector<edge_t> recv_data(local_degrees.size(), handle.get_stream());
+  if (col_rank == 0) {
+    thrust::fill(handle.get_thrust_policy(), recv_data.begin(), recv_data.end(), edge_t{0});
+  }
+  for (int i = 0; i < col_size - 1; ++i) {
+    if (col_rank == i) {
+      comm.device_send(
+        temp_input.begin(), temp_input.size(), comm_rank + row_size, handle.get_stream());
+    }
+    if (col_rank == i + 1) {
+      comm.device_recv(
+        recv_data.begin(), recv_data.size(), comm_rank - row_size, handle.get_stream());
+      thrust::transform(handle.get_thrust_policy(),
+                        temp_input.begin(),
+                        temp_input.end(),
+                        recv_data.begin(),
+                        temp_input.begin(),
+                        thrust::plus<edge_t>());
+    }
+    col_comm.barrier();
+  }
+  // Get global degrees
+  device_bcast(col_comm,
+               temp_input.begin(),
+               temp_input.begin(),
+               temp_input.size(),
+               col_size - 1,
+               handle.get_stream());
+
+  return std::make_tuple(std::move(recv_data), std::move(temp_input));
```

Comment:
Note that sizes of recv_data & temp_input are V/P * col_comm_size, so this can be a scalability bottleneck (in terms of memory capacity especially when col_comm_size > E/V).

And should we really store the major_degrees and exclusive_scan of the local major degrees for the entire major range?

I am not sure how the sampling will be implemented end-to-end, but if I make a guess.

1) store major_degrees (or major weight sums) for V/P vertices (not for the V/ P * col_comm_size) in each GPU.

2) Once the active source vertices are given, shuffle sources to the corresponding vertex partition, then, based on the global degrees (or weight sums), generate random numbers within the ranges.

Broadcast the random numbers (only for the active sources) to the GPUs in the same column communicator. Compute exclusive scan of the local degrees "only for the active source vertices". Based on the random numbers, sample edges. Then, collect edges back to the final destination GPUs.

This may involve some recomputing (but only for the active source vertices) but memory footprint will be much more scalable. And recomputing cost may be pretty moderate when # active sources << V.

Reply:
> And recomputing cost may be pretty moderate when # active sources << V.

I am not sure about this. For the first call this might be true. But the subsequent calls will probably increase the number of active vertices.

As per my current understanding, my aim was to calculate these values (degrees and offsets) once because the number of gather calls could be quite large.
Once I get a sense of what the memory vs communication trade off looks like I will fix this.

See [here](https://github.com/rapidsai/cugraph/pull/2064/files/258b7d4c1e4a185c1b454dd8590f0031dd61496f#r804779751).